measureme/src/counters.rs

```diff
@@ -0,0 +1,869 @@
+}
+
+impl Counter {
+    pub fn by_name(name: &str) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        Ok(match name {
+            WallTime::NAME => Counter::WallTime(WallTime::new()),
+            Instructions::NAME => Counter::Instructions(Instructions::new()?),
+            InstructionsMinusIrqs::NAME => {
+                Counter::InstructionsMinusIrqs(InstructionsMinusIrqs::new()?)
+            }
+            InstructionsMinusRaw0420::NAME => {
+                Counter::InstructionsMinusRaw0420(InstructionsMinusRaw0420::new()?)
+            }
+            _ => return Err(format!("{:?} is not a valid counter name", name).into()),
+        })
+    }
+
+    pub(super) fn describe_as_json(&self) -> String {
+        let (name, units) = match self {
+            Counter::WallTime(_) => (
+                WallTime::NAME,
+                r#"[["ns", 1], ["μs", 1000], ["ms", 1000000], ["s", 1000000000]]"#,
+            ),
+            Counter::Instructions(_) => (Instructions::NAME, r#"[["instructions", 1]]"#),
+            Counter::InstructionsMinusIrqs(_) => {
+                (InstructionsMinusIrqs::NAME, r#"[["instructions", 1]]"#)
+            }
+            Counter::InstructionsMinusRaw0420(_) => {
+                (InstructionsMinusRaw0420::NAME, r#"[["instructions", 1]]"#)
+            }
+        };
+        format!(r#"{{ "name": "{}", "units": {} }}"#, name, units)
+    }
+
+    #[inline]
+    pub(super) fn since_start(&self) -> u64 {
+        match self {
+            Counter::WallTime(counter) => counter.since_start(),
+            Counter::Instructions(counter) => counter.since_start(),
+            Counter::InstructionsMinusIrqs(counter) => counter.since_start(),
+            Counter::InstructionsMinusRaw0420(counter) => counter.since_start(),
+        }
+    }
+}
+
+pub struct WallTime {
+    start: Instant,
+}
+
+impl WallTime {
+    const NAME: &'static str = "wall-time";
+
+    pub fn new() -> Self {
+        WallTime {
+            start: Instant::now(),
+        }
+    }
+
+    #[inline]
+    fn since_start(&self) -> u64 {
+        self.start.elapsed().as_nanos() as u64
+    }
+}
+
+pub struct Instructions {
+    instructions: hw::Counter,
+    start: u64,
+}
+
+impl Instructions {
+    const NAME: &'static str = "instructions:u";
+
+    pub fn new() -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let model = hw::CpuModel::detect()?;
+        let instructions = hw::Counter::new(&model, HwCounterType::Instructions)?;
+        let start = instructions.read();
+        Ok(Instructions {
+            instructions,
+            start,
+        })
+    }
+
+    #[inline]
+    fn since_start(&self) -> u64 {
+        self.instructions.read().wrapping_sub(self.start)
+    }
+}
+
+pub struct InstructionsMinusIrqs {
+    instructions: hw::Counter,
+    irqs: hw::Counter,
+    start: u64,
+}
+
+impl InstructionsMinusIrqs {
+    const NAME: &'static str = "instructions-minus-irqs:u";
```

Comment:
Maybe use `instructions:u` as name here for the real amount of userspace instructions and use something like `instructions:u+int` as name for the amount of userspace instructions combined with any bogus counted instructions caused by interrupt returns or similar things.

Reply:
If we do that I'd be against using `instructions:u` as the name for anything, because `perf stat` *does not* do any subtracting, and we used the `instructions:u` name specifically for `perf stat` familiarity (which is why `perf.rust-lang.org` also uses it).

Furthermore, we might want to try and get `perf.rust-lang.org` to record the right counter and *also* show `instructions-minus-irqs:u`, so ideally we'd use different names to tell them apart.

But I agree the current setup is not the most user-friendly, it was just the easiest way to keep things clear while we worked on this.